Backup protected status can vary per location

mdegat01 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

Backups can now be easily encrypted and decrypted while streaming and core does this regularly. Therefore two backups that are otherwise identical except one is encrypted and one is not need to be considered the same. And Supervisor should tell clients the protected status of backups across locations.

## Type of change

<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (which adds functionality to the supervisor)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:
- Link to cli pull request:
- Link to client library pull request:

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Ruff (`ruff format supervisor tests`)
- [ ] Tests have been added to verify that the new code works.

If API endpoints or add-on configuration are added/changed:

- [ ] Documentation added/updated for [developers.home-assistant.io][docs-repository]
- [ ] [CLI][cli-repository] updated (if necessary)
- [ ] [Client library][client-library-repository] updated (if necessary)

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[docs-repository]: https://github.com/home-assistant/developers.home-assistant
[cli-repository]: https://github.com/home-assistant/cli
[client-library-repository]: https://github.com/home-assistant-libs/python-supervisor-client/


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Enhanced backup location management with detailed protection status.
	- Added ability to track protected backup locations.

- **Improvements**
	- Improved backup information retrieval.
	- More robust handling of backup file paths and encryption status.
	- Centralized password validation for backup locations.

- **Bug Fixes**
	- Updated backup location data structure for more accurate tracking.

- **Tests**
	- Added new test cases for backup encryption and protection scenarios.
	- Enhanced existing tests to validate new backup location structure.

These changes provide a more comprehensive approach to managing backup locations, ensuring detailed information about backup protection and file paths.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

agners commented:
The `tests/backups/test_manager.py::test_backup_remove_error` test failure actually uncovers a bug in that test in general: The backup hasn't been in the `test` location at all, this test should have failed even before this change :see_no_evil: .

I've updated the test to actually copy the backup to that mount location as well, and fixed the test accordingly.

agners commented on supervisor/backups/backup.py:
```diff
@@ -252,7 +258,27 @@ def data(self) -> dict[str, Any]:
+
+        # Compare all fields except ones about protection. Current encryption status does not affect equality
+        keys = self._data.keys() | other._data.keys()
+        for k in keys - {ATTR_PROTECTED, ATTR_CRYPTO}:
```
Tests with an unencrypted/encrypted backup also made the `registries` setting to light up:

```
2025-01-23 11:56:19.845 DEBUG (MainThread) [supervisor.backups.backup] Backup 5ba8f1d7 and 5ba8f1d7 not equal because docker field has different value: {'registries': {'hub.docker.com': {'username': 'agners', 'password': '/aDMOAXyCd5fG1ZR2j4nHA=='}}} and {'registries': {'hub.docker.com': {'username': 'agners', 'password': 'gaga'}}}
2025-01-23 11:56:19.845 ERROR (MainThread) [supervisor.backups.manager] Ignoring backup 5ba8f1d7 in backup_test due to: Backup in backup_test and None both have slug 5ba8f1d7 but are not the same!
```

Technically we could dive into the field and only ignore the password (since this is the only part which is encrypted/unencrypted). However, the idea is to move the settings into a Supervisor specific tarball anyways, so I'd say we just ignore the whole attribute for now:

```suggestion
        for k in keys - {ATTR_PROTECTED, ATTR_CRYPTO, ATTR_DOCKER}:
```

emontnemery commented on supervisor/backups/backup.py:
```diff
@@ -263,7 +263,7 @@ def __eq__(self, other: Any) -> bool:
         # Compare all fields except ones about protection. Current encryption status does not affect equality
         keys = self._data.keys() | other._data.keys()
-        for k in keys - {ATTR_PROTECTED, ATTR_CRYPTO}:
+        for k in keys - {ATTR_PROTECTED, ATTR_CRYPTO, ATTR_DOCKER}:
```
Do we have tests to ensure these attributes are ignored?

mdegat01 replied:
There's a test to confirm an encrypted backup and unencrypted backup can be consolidated, which means those fields were ignored. I didn't have `docker` in there before so my backup fixtures didn't have that, I'll have to check on that.

Technically for docker we should only be ignoring it if one backup is encrypted and one isn't, otherwise it should be the same. I don't know if we want to deal with that for a field to be deprecated though

agners replied:
> I don't know if we want to deal with that for a field to be deprecated though

Yeah let's not bother about that.

But adding `docker` attributes and make sure they get ignored in the test make sense :+1: .

agners replied:
Ok, I've updated the test fixture files to include a docker repository setting. The test fails without the change above, but with the change merged succeeds :+1: .

emontnemery commented on tests/backups/test_manager.py:
```diff
@@ -1734,12 +1734,20 @@ async def test_backup_remove_error(
     healthy_expected: bool,
 ):
     """Test removing a backup error."""
```
Why did this test start failing?

agners replied:
Essentially, because one of the backup didn't exist before. But previously the test just "created" the location, and now because `all_locations` it is a dictionary in a dictionary that lead to an error.

A bit more detail in the comment below at https://github.com/home-assistant/supervisor/pull/5569#issuecomment-2609783007.

emontnemery commented on tests/backups/test_manager.py:
```diff
@@ -1734,12 +1734,20 @@ async def test_backup_remove_error(
-    copy(get_fixture_path("backup_example.tar"), coresys.config.path_backup)
-    await coresys.backups.reload(location=None, filename="backup_example.tar")
+    backup_file = get_fixture_path("backup_example.tar")
+    if location_name is None:
+        copy(backup_file, coresys.config.path_backup)
+        location = None
+    else:
+        (mount_dir := coresys.config.path_mounts / location_name).mkdir()
+        copy(backup_file, mount_dir)
+        location = coresys.mounts.get(location_name)
```
Instead of the if-statement, consider adding additional parametrization variable, for the subdirectory and maybe the key to get the location from `coresys.mounts.get`

agners replied:
Unfortunately how to get the location is different for None and mount backups. I've reused existing code now, so that makes the if statement unnecessary.

mdegat01 commented:
> The `tests/backups/test_manager.py::test_backup_remove_error` test failure actually uncovers a bug in that test in general: The backup hasn't been in the `test` location at all, this test should have failed even before this change 🙈 .
> 
> I've updated the test to actually copy the backup to that mount location as well, and fixed the test accordingly.

Wait what was the failure? I didn't think the file not being in that location would cause a failure, I'm pretty sure other tests do that too?
